helix-view/src/editor.rs

```diff
@@ -398,18 +409,64 @@ impl Editor {
             .selections
             .remove(&id);
 
-        if close_buffer {
-            // get around borrowck issues
-            let doc = &self.documents[&view.doc];
+        self.tree.remove(id);
+        self._refresh();
+    }
 
-            if let Some(language_server) = doc.language_server() {
-                tokio::spawn(language_server.text_document_did_close(doc.identifier()));
-            }
-            self.documents.remove(&view.doc);
+    pub fn close_document(&mut self, doc_id: DocumentId, force: bool) -> anyhow::Result<()> {
+        let doc = match self.documents.get(&doc_id) {
+            Some(doc) => doc,
+            None => anyhow::bail!("document does not exist"),
+        };
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
+        }
+
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        let views_to_close = self
+            .tree
+            .views()
+            .filter_map(|(view, focus)| {
+                if view.doc == doc_id {
+                    Some((view.id, focus))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for (view_id, _focus) in views_to_close {
+            self.close(view_id);
+        }
+
+        self.documents.remove(&doc_id);
+
+        if self.tree.views().next().is_none() {
+            let doc_id = self
+                .documents
+                .iter()
+                .map(|(&doc_id, _)| doc_id)
+                .next()
+                .unwrap_or_else(|| self.new_document(Document::default()));
+            let view = View::new(doc_id);
+            let view_id = self.tree.insert(view);
+            let view = self.tree.get(view_id);
+            let doc = self.documents.get_mut(&view.doc).unwrap();
```

Comment:
```suggestion
            let doc = self.documents.get_mut(doc_id).unwrap();
```
Can't we just do this?